### PR TITLE
Center Based Algorithms

### PR DESCRIPTION
**Pull request by Rama-Has**

### CODE REVIEW

File: src/utils/center_based_clustering_algorithms.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+        else:
+            model_instance = model(n_clusters = k, random_state = 42)
+
+        #fit the model and get the predicted lable for each sample
+        cluster_labels = model_instance.fit_predict(data)
+
+        #calculate the silhouette score.
+        silhouette_avg = silhouette_score(data, cluster_labels)
+
+        #append the result of the average silhouette score
```
Comment (mawada-sweis) [inline src/utils/center_based_clustering_algorithms.py:40]: Did you mean that you will append the result to the silhouette list?

File: src/utils/center_based_clustering_algorithms.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+        model_name(str): name of the used model that generated the silhouette score values.
+    """ 
+    plt.plot(potential_k, silhouette_score_for_ks)
+    plt.xlabel('K') 
+    plt.ylabel('Silhouette score')   
+    plt.title('Silhouette Analysis For Optimal k Using ' + model_name)
+    plt.show()
+             
+
+def plot_silhouette(n_clusters, model, data) -> None:
```
Comment (mawada-sweis) [inline src/utils/center_based_clustering_algorithms.py:61]: Can you please add the type hint to these arguments?

File: src/utils/center_based_clustering_algorithms.py [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+    # Fit the data to the visualizer
+    visualizer.fit(data)
+
+    # Finalize and render the figure 
+    visualizer.show() 
+
+
+def plot_elbow(data: pd.DataFrame, model, potential_k_list: list) -> None:
+    """Plot the elbow method to ffind the best value for k for a given clustering model.
```
Comment (mawada-sweis) [inline src/utils/center_based_clustering_algorithms.py:87]: Can you please re-write this description?

File: src/utils/center_based_clustering_algorithms.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    K_means_visualizer_reduced_data.show() 
+    
+
+def train_model(model, n_clusters: int, data: pd.DataFrame): 
+    """Train a clustering model using the given data with the given number of clusters.
+
+    Args:
+        model: a clustering model reference.
+        n_clusters(int): number of clusters to be used in the training process.
+        data(pd.DataFrame): the dataframe taht contains data to be used in the training preocess.
```
Comment (mawada-sweis) [inline src/utils/center_based_clustering_algorithms.py:112]: I think you mean 'that', right?

File: src/utils/center_based_clustering_algorithms.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    K_means_visualizer_reduced_data = KElbowVisualizer(model, k=potential_k_list, timings=False)
+
+    # Fit the data to the visualizer
+    K_means_visualizer_reduced_data.fit(data)
+
+    # Finalize and render the figure 
+    K_means_visualizer_reduced_data.show() 
+    
+
+def train_model(model, n_clusters: int, data: pd.DataFrame):
```
Comment (mawada-sweis) [inline src/utils/center_based_clustering_algorithms.py:106]: The type hint for this case is a tuple of ndarray

File: src/utils/center_based_clustering_algorithms.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+    centers = model_instance.cluster_centers_
+
+    return labels, centers
+
+
+def plot_2d_scatter(data: pd.DataFrame, x_axis_column_name: str, y_axis_column_name: str, labels: np.ndarray, embedded_centers: np.ndarray) -> None: 
+    """Plot a 2d scatter plot of the given data each sample will be colored based on its cluster label.
+
+    Args:
+        data(pd.DataFrame): the dataframe taht contains data to be used in x-axis and y-axis.
```
Comment (mawada-sweis) [inline src/utils/center_based_clustering_algorithms.py:137]: Again, I think you mean 'that'...

File: src/utils/center_based_clustering_algorithms.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    data=data, 
+    hue=labels,  
+    palette="Set2"
+    )
+
+    # Plot black dots at the values of each embedded center for clusters.
+    plt.scatter(embedded_centers[:, 0], embedded_centers[:, 1], c='black', s=200, alpha=0.5)
+
+    #change the plot title 
+    scat.set_title( "Clustering results from .... data")
```
Comment (mawada-sweis) [inline src/utils/center_based_clustering_algorithms.py:162]: Why is the title written like this?

**Review by mawada-sweis — COMMENTED** [review]
I will review the notebook in separate review... Thank you for your effort, the code is excellent.

**Review by zubaidasader — APPROVED** [review]

**Review by eleenkmail — APPROVED** [review]

**Review by mawada-sweis — COMMENTED** [review]
Great.

**Review by mawada-sweis — APPROVED** [review]

**Pull request merged**